etcdserver/server.go

```diff
@@ -809,36 +809,39 @@ func (s *EtcdServer) applyConfChange(cc raftpb.ConfChange, confState *raftpb.Con
 	return false, nil
 }
 
-// TODO: non-blocking snapshot
 func (s *EtcdServer) snapshot(snapi uint64, confState *raftpb.ConfState) {
-	d, err := s.store.Save()
-	// TODO: current store will never fail to do a snapshot
-	// what should we do if the store might fail?
-	if err != nil {
-		log.Panicf("etcdserver: store save should never fail: %v", err)
-	}
-	err = s.r.raftStorage.Compact(snapi, confState, d)
-	if err != nil {
-		// the snapshot was done asynchronously with the progress of raft.
-		// raft might have already got a newer snapshot and called compact.
-		if err == raft.ErrCompacted {
-			return
+	clone := s.store.Clone()
+
+	go func() {
+		d, err := clone.SaveNoCopy()
+		// TODO: current store will never fail to do a snapshot
+		// what should we do if the store might fail?
+		if err != nil {
+			log.Panicf("etcdserver: store save should never fail: %v", err)
 		}
-		log.Panicf("etcdserver: unexpected compaction error %v", err)
-	}
-	log.Printf("etcdserver: compacted log at index %d", snapi)
+		err = s.r.raftStorage.Compact(snapi, confState, d)
```

Comment:
@barakmich snapshot in raft is an array of bytes. attempting to change snapshot to io.reader is pre-mature until we "fix" raft snapshot.